apps/rewards/src/invite/rewardSender.service.ts

```diff
@@ -0,0 +1,156 @@
+import { KomenciLoggerService } from '@app/komenci-logger'
+import { ContractKit } from '@celo/contractkit'
+import { Inject, Injectable } from '@nestjs/common'
+import { Cron, CronExpression } from '@nestjs/schedule'
+import BigNumber from 'bignumber.js'
+import { EntityManager, In, Raw, Repository } from 'typeorm'
+import { appConfig, AppConfig } from '../config/app.config'
+import { RelayerProxyService } from '../relayer/relayer_proxy.service'
+import { InviteReward, RewardStatus } from './inviteReward.entity'
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEI_PER_UNIT = 1000000000000000000
+
+interface WatchedInvite {
+  inviteId: string
+  txHash: string
+  sentAt: number
+  markAsDeadLetterIfFailed: boolean
+}
+
+@Injectable()
+export class RewardSenderService {
+  watchedInvites: Set<WatchedInvite>
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly relayerProxyService: RelayerProxyService,
+    @Inject(appConfig.KEY)
+    private readonly appCfg: AppConfig,
+    private readonly contractKit: ContractKit,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.watchedInvites = new Set()
+  }
+
+  async wasTxSentSuccesfully(txHash: string) {
+    const tx = await this.contractKit.web3.eth.getTransaction(txHash)
+    return Boolean(tx.blockHash)
+  }
+
+  @Cron(CronExpression.EVERY_SECOND)
+  async checkWatchedInvites() {
+    await Promise.all(
+      [...this.watchedInvites].map(async invite => {
+        if (await this.wasTxSentSuccesfully(invite.txHash)) {
+          await this.inviteRewardRepository.update(invite.inviteId, {
+            state: RewardStatus.Completed
+          })
+          this.watchedInvites.delete(invite)
+        } else if (
+          Date.now() - invite.sentAt >
+          this.appCfg.transactionTimeoutMs
+        ) {
+          await this.inviteRewardRepository.update(invite.inviteId, {
+            state: invite.markAsDeadLetterIfFailed
+              ? RewardStatus.DeadLettered
+              : RewardStatus.Failed
+          })
+          this.watchedInvites.delete(invite)
+        }
+      })
+    )
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async checkFailedInvites() {
+    await this.inviteRewardRepository.manager.transaction(
+      async (entityManager: EntityManager) => {
+        const repository = entityManager.getRepository<InviteReward>(
+          'invite_reward'
+        )
+        const invites = await repository
+          .createQueryBuilder()
+          .setLock('pessimistic_read')
+          .where({
+            state: In([RewardStatus.Submitted, RewardStatus.Failed]),
+            createdAt: Raw(
```

Comment:
I'm not sure how efficient this query will be. Do you know if this range query will leverage an index? Should we specify a max look back timestamp (ex. 2 hours)? Should we specify an upper limit on the result count?

Reply:
Well, under normal circumstances there should never be many results in this query. 

Rewards should never end up in `Submitted` state for more than 30 minutes, if that happens I'm assuming that the transaction failed, maybe there's a way to do a more explicit check for that or add a timeout to the transaction? Maybe I'm not handling that case gracefully enough.

As for failed transactions, this cron job is running every minute and checks failed transactions from 30 minutes ago, so the number of failed transactions here are the ones that failed in a 1 minute window and they are marked as `DeadLettered` after that, so there should never be a Failed transaction from more than 2 hours ago. I guess this could be off if the server is down and then a bunch of invites are sent at the same time. I can add a limit of 100 elements here so that if that happens invite sending is spread out over several runs of this cron job.

I don't think this is using an index,  are you worried that over time this will end up being a very slow query? It seems unlikely to be a bottleneck in the short term, but I can appreciate trying to future-proof this. Maybe we can add a metric to track this and throw an alarm if it's taking more than X time to complete?

Reply:
Yeah, I'm just worried that if we have a long history of records, even though most of them will be settled (Success | Deadlettered) that it will still search through all of them each time. If we're indexing on `state` column it may not be a problem.